### PR TITLE
Refactor entry point finder code

### PR DESCRIPTION
**Pull request by jmarolf**
## Description

fixes https://github.com/dotnet/roslyn/issues/60696

I found that we have  a few different places that look for entry points and they all use different logic and all have different bugs. The intent here is to unify the logic to find the entry points in a given compilation and check whether a method symbol is a valid entry point.

## Change List

1. [77e2281](https://github.com/dotnet/roslyn/pull/60614/commits/77e22811f22744d2c6b5e48631800510612bbdd5): Refactors our `IEntryPointFinderService` to accept a compilation instead of a namespace symbol
1. [5c3c04a](https://github.com/dotnet/roslyn/pull/60614/commits/5c3c04a667e7f53eeb766b02f2a51e5e5c39a3d0): Fixes some bugs in how we discover entry points 
1. [e80698f](https://github.com/dotnet/roslyn/pull/60614/commits/e80698fab960905d5ecf9777f283ccb13a717a0e): Adds some helper methods to make finding entry points less verbose 
1. [b389bbd](https://github.com/dotnet/roslyn/pull/60614/commits/b389bbdfd0c9a4afefb0a90916e3b7fed6623f56): Consolidates all the places in Roslyn that searches for possible entry points to use the same code paths 
1. [148e0a8](https://github.com/dotnet/roslyn/pull/60614/commits/148e0a87d287b81b36ec2f8f67ddb0178c26223b): For files that were touched that have not yet been nullable annotated, do so.

### No public API changes

Though it might be worth talking to the compiler team about adding something for this. The compiler has internal apis that can compute this.

### New API for first party consumers (Project System)

```diff
namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
{
    internal interface IEntryPointFinderService : ILanguageService
    {
        /// <summary>
        /// Finds the types that contain entry points like the Main method in a give namespace.
        /// </summary>
        /// <param name="symbol">The namespace to search.</param>
        /// <param name="findFormsOnly">Restrict the search to only Windows Forms classes. Note that this is only implemented for VisualBasic</param>
+       [Obsolete("FindEntryPoints on a INamespaceSymbol is deprecated, please pass in the Compilation instead.")]
        IEnumerable<INamedTypeSymbol> FindEntryPoints(INamespaceSymbol symbol, bool findFormsOnly);

+       /// <summary>
+       /// Finds the types that contain entry points like the Main method in a given compilation.
+       /// </summary>
+       /// <param name="compilation">The compilation to search.</param>
+       /// <param name="findFormsOnly">Restrict the search to only Windows Forms classes. Note that this is only implemented for VisualBasic</param>
+       IEnumerable<INamedTypeSymbol> FindEntryPoints(Compilation compilation, bool findFormsOnly);
    }
}
```
**NOTE**: the new api added in [77e2281](https://github.com/dotnet/roslyn/pull/60614/commits/77e22811f22744d2c6b5e48631800510612bbdd5) also marks `FindEntryPoints(INamespaceSymbol symbol, bool findFormsOnly)`. as `Obsolete`. Once this is merged and all upstream consumers move to the new api in https://github.com/dotnet/project-system/pull/8044 the plan is to delete the obsolete method.

### New Internal IsEntryPoint Api for use in roslyn anlayzers

```diff
namespace Microsoft.CodeAnalysis.Shared.Extensions
{
    internal static partial class IMethodSymbolExtensions
    {
-       public static bool IsEntryPoint(this IMethodSymbol methodSymbol, INamedTypeSymbol? taskType, INamedTypeSymbol? genericTaskType);
+       public static bool IsEntryPoint(this IMethodSymbol methodSymbol, StringComparer languageStringComparer, ImmutableArray<string> entryPointMethodNames, INamedTypeSymbol? taskType = null, INamedTypeSymbol? genericTaskType = null);
    }
}


namespace Microsoft.CodeAnalysis.CSharp.Shared.Extensions
{
    internal static class IMethodSymbolExtensions
    {
+        public static bool IsCSharpEntryPoint(this IMethodSymbol methodSymbol, INamedTypeSymbol? taskType, INamedTypeSymbol? genericTaskType);
    }
}

namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
{
    internal static class IMethodSymbolExtensions
    {
+        public static bool IsVisualBasicEntryPoint(this IMethodSymbol methodSymbol);
    }
}
```

## ~~TODO~~

(tests added in [4821422](https://github.com/dotnet/roslyn/pull/60614/commits/48214225624e9b99072ef72739a0fa6096cc143e))
- [x] Write new tests for `IsEntryPoint`, `IsCSharpEntryPoint`, and `IsVisualBasicEntryPoint` extension methods
- [x] Write new tests for `Microsoft.VisualStudio.LanguageServices.CSharp.ProjectSystemShim.EntryPointFinder`
- [x] Write new tests for `Microsoft.VisualStudio.LanguageServices.CSharp.ProjectSystemShim.CSharpEntryPointFinderService `
- [x] Write new tests for `Microsoft.VisualStudio.LanguageServices.VisualBasic.ProjectSystemShim.EntryPointFinder`
- [x] Write new tests for `Microsoft.VisualStudio.LanguageServices.VisualBasic.ProjectSystemShim.VisualBasicEntryPointFinderService`

### CODE REVIEW

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 1]
```diff
@@ -418,6 +418,22 @@ public static ImmutableArray<T> WhereAsArray<T>(this ImmutableArray<T> array, Fu
...
+            int n = array.Length;
+            for (int i = 0; i < n; i++)
+            {
+                var a = array[i];
+
+                if (predicate(a, arg1, arg2))
+                {
+                    return true;
+                }
+            }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:432]:
```suggestion
        {
            for (var value in array)
            {
                if (predicate(value, arg1, arg2))
                    return true;
            }
```

File: src/VisualStudio/CSharp/Impl/ProjectSystemShim/EntryPointFinder.cs [hunk 2]
```diff
@@ -2,26 +2,46 @@
...
 using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
 
 namespace Microsoft.VisualStudio.LanguageServices.CSharp.ProjectSystemShim
 {
     internal class EntryPointFinder : AbstractEntryPointFinder
     {
-        protected override bool MatchesMainMethodName(string name)
-            => name == "Main";
+        private readonly INamedTypeSymbol? _task;
+        private readonly INamedTypeSymbol? _taskOf;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/CSharp/Impl/ProjectSystemShim/EntryPointFinder.cs:17]: _taskOfTType

File: src/VisualStudio/CSharp/Impl/ProjectSystemShim/EntryPointFinder.cs [hunk 3]
```diff
@@ -2,26 +2,46 @@
...
+using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
 
 namespace Microsoft.VisualStudio.LanguageServices.CSharp.ProjectSystemShim
 {
     internal class EntryPointFinder : AbstractEntryPointFinder
     {
-        protected override bool MatchesMainMethodName(string name)
-            => name == "Main";
+        private readonly INamedTypeSymbol? _task;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/CSharp/Impl/ProjectSystemShim/EntryPointFinder.cs:16]:
```suggestion
        private readonly INamedTypeSymbol? _taskType;
```

File: src/VisualStudio/CSharp/Impl/ProjectSystemShim/EntryPointFinder.cs [hunk 4]
```diff
@@ -2,26 +2,46 @@
...
+        private readonly INamedTypeSymbol? _taskOf;
 
+        public EntryPointFinder(Compilation? compilation)
+        {
+            _task = compilation?.TaskType();
+            _taskOf = compilation?.TaskOfTType();
+        }
+
+        [Obsolete("FindEntryPoints on a INamespaceSymbol is deprecated, please pass in the Compilation instead.")]
         public static IEnumerable<INamedTypeSymbol> FindEntryPoints(INamespaceSymbol symbol)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/CSharp/Impl/ProjectSystemShim/EntryPointFinder.cs:26]: why not just remove these?

File: src/VisualStudio/CSharp/Impl/ProjectSystemShim/EntryPointFinder.cs [hunk 5]
```diff
@@ -2,26 +2,46 @@
...
-            var visitor = new EntryPointFinder();
+            var visitor = new EntryPointFinder(symbol.ContainingCompilation);
             visitor.Visit(symbol);
             return visitor.EntryPoints;
         }
+
+        public static IEnumerable<INamedTypeSymbol> FindEntryPoints(Compilation compilation)
+        {
+            var visitor = new EntryPointFinder(compilation);
+            visitor.Visit(compilation.SourceModule.GlobalNamespace);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/CSharp/Impl/ProjectSystemShim/EntryPointFinder.cs:40]: i'm not sure what the value of hte compilation is.  we're just using it to get the Global namespace.  So what was wrong with the prior form?

or, if we do want to mandate that you only use the Compilation, we should remove the namespace version.

File: src/VisualStudio/Core/Def/ProjectSystem/IEntryPointFinderService.cs [hunk 6]
```diff
@@ -17,6 +16,14 @@ internal interface IEntryPointFinderService : ILanguageService
         /// </summary>
         /// <param name="symbol">The namespace to search.</param>
         /// <param name="findFormsOnly">Restrict the search to only Windows Forms classes. Note that this is only implemented for VisualBasic</param>
+        [Obsolete("FindEntryPoints on a INamespaceSymbol is deprecated, please pass in the Compilation instead.")]
         IEnumerable<INamedTypeSymbol> FindEntryPoints(INamespaceSymbol symbol, bool findFormsOnly);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/ProjectSystem/IEntryPointFinderService.cs:20]: why not remove this?

File: src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb [hunk 7]
```diff
@@ -15,14 +16,15 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.ProjectSystemShim
             Me._findFormsOnly = findFormsOnly
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb:16]: i'm not really understanding the purpose of this flag.

**Comment by CyrusNajmabadi** [comment]
> I found that we have a few different places that look for entry points and they all use different logic and all have different bugs. The intent here is to unify the logic to find the entry points in a given compilation and check whether a method symbol is a valid entry point.

Can we just propose an API for the compiler layer and just actually return whatever value they return?

File: src/VisualStudio/CSharp/Impl/ProjectSystemShim/EntryPointFinder.cs (hunk 4 above)
Comment (jmarolf) [reply src/VisualStudio/CSharp/Impl/ProjectSystemShim/EntryPointFinder.cs:26]: see comment in the PR description: 

> NOTE: the new api added in [77e2281](https://github.com/dotnet/roslyn/pull/60614/commits/77e22811f22744d2c6b5e48631800510612bbdd5) also marks FindEntryPoints(INamespaceSymbol symbol, bool findFormsOnly). as Obsolete. Once this is merged and all upstream consumers move to the new api in https://github.com/dotnet/project-system/pull/8044 the plan is to delete the obsolete method.

Don't want to break project system so this represents an intermediate state. the Obsolete method will be removed in a follow up PR as soon as the project system PR is merged.

File: src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb (hunk 7 above)
Comment (jmarolf) [reply src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/EntryPointFinder.vb:16]: the Visual Basic project system has a feature that allows you set any Form as the entry point. This was to enable more rapid development for winforms developers. The entry point service only discovers winforms types in a compilation when in this mode.

**Comment by jmarolf** [comment]
> > I found that we have a few different places that look for entry points and they all use different logic and all have different bugs. The intent here is to unify the logic to find the entry points in a given compilation and check whether a method symbol is a valid entry point.
> 
> Can we just propose an API for the compiler layer and just actually return whatever value they return?

Not opposed to this. My assumption is that we would want something like this

```diff
namespace Microsoft.CodeAnalysis
{
    public abstract Compilation
    {
+       /// <summary>
+       /// Determines whether the given 
+       /// <see cref="IMethodSymbol ">IMethodSymbol</see> is a valide entry point for the compilation
+       /// </summary>
+       bool IsEntryPoint(IMethodSymbol methodSymbol);
    }
}
```

Let me open up a proposal

**Pull request closed without merging**